### PR TITLE
Add INS Implementation

### PR DESCRIPTION
**Pull request by PhyrexTsai**
## IIP-6 INS Implementation

This is a **DRAFT** version of the INS mechanism.

Implementation:
- INS Registry
- INS Resolver
- INS Registrar

### CODE REVIEW

File: IIPS/iip-6.md [hunk 1]
```diff
@@ -143,6 +143,240 @@ Returns the hash for a given address's reverse records.
...
+    @eventlog(indexed=3)
+    def NewOwner(self, node: bytes, label: bytes, owner: Address) : pass
+
+    @eventlog(indexed=2)
+    def NewResolver(self, node: bytes, resolver: Address) : pass
+
+    @eventlog(indexed=2)
+    def NewTTL(self, node: bytes, ttl: int) : pass
+
+    def __init__(self, db: IconScoreDatabase, addr_owner: Address) -> None:
```
Comment (sink772) [inline IIPS/iip-6.md]: `addr_owner` has been removed from `__init__` function definition.  Please remove the param.

File: IIPS/iip-6.md [hunk 2]
```diff
@@ -143,6 +143,240 @@ Returns the hash for a given address's reverse records.
...
+    @eventlog(indexed=3)
+    def MultihashChanged(node: bytes, name: str, multihash: bytes) : pass
+
+    @eventlog(indexed=3)
+    def TextChanged(node: bytes, key: str, value: str) : pass
+
+    @eventlog(indexed=2)
+    def NameChanged(node: bytes, name: str) : pass
+
+    def __init__(self, db: IconScoreDatabase, owner: Address) -> None:
```
Comment (sink772) [inline IIPS/iip-6.md]: Same as above.

File: IIPS/iip-6.md [hunk 3]
```diff
@@ -143,6 +143,240 @@ Returns the hash for a given address's reverse records.
...
+    @interface
+    def setSubnodeOwner(node: bytes, label: bytes, owner: Address) : pass
+
+
+class INSRegistrar(IconScoreBase):
+
+    __ADDR_INS_REGISTRY_SCORE = 'addr_ins_registry_score'
+    __ROOT_NODE = 'root_node'
+
+    def __init__(self, db: IconScoreDatabase, owner: Address) -> None:
```
Comment (sink772) [inline IIPS/iip-6.md]: Same as above.

File: IIPS/iip-6.md [hunk 4]
```diff
@@ -143,6 +143,240 @@ Returns the hash for a given address's reverse records.
...
+    def NewOwner(self, node: bytes, label: bytes, owner: Address) : pass
+
+    @eventlog(indexed=2)
+    def NewResolver(self, node: bytes, resolver: Address) : pass
+
+    @eventlog(indexed=2)
+    def NewTTL(self, node: bytes, ttl: int) : pass
+
+    def __init__(self, db: IconScoreDatabase, addr_owner: Address) -> None:
+        super().__init__(db, addr_owner)
```
Comment (sink772) [inline IIPS/iip-6.md]: You should simply call `super().__init__(db)` here.

**Comment by PhyrexTsai** [comment]
@sink772 Thank you for your advise and help, I have put the INS source code to the repository of https://github.com/PortalNetwork/ins

**Comment by sink772** [comment]
@PhyrexTsai Sounds great! Then please update the draft document to include the link of your repository. Once you update the document, I will merge your changes.

**Comment by PhyrexTsai** [comment]
@sink772 I've update the document, and add the link to INS source code, thank you.

**Pull request merged**